### PR TITLE
[haproxy] Redis Config + Metrics

### PR DESCRIPTION
**Pull request by splattner**
Signed-off-by: Sebastain Plattner <sebastian.plattner@gmail.com>

#### What this PR does / why we need it:

Extend the chart to allow a Redis configuration with dns autodiscovery of the master node
Extend the chart to allow adding of extra sidecards into the haproxy deployment
Extend the chart to enable metrics and Prometheus ServiceMonitor and PrometheusRule for all configurations

#### Checklist
- [x] [DCO](https://github.com/appuio/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR contains starts with chart name e.g. `[chart]`

### CODE REVIEW

File: haproxy/README.md [hunk 1]
```diff
@@ -48,11 +48,35 @@ create your own configmap and values.
...
+| `service.port` | 	Kubernetes Service port (haproxy) | `30636`
+| `service.type` | Kubernetes Service type (haproxy) | `ClusterIP`
+| `service.additionalPorts` | additional Ports to be added to the haproxy service (e.g. for sidecars) | `[]`
+| `sidecarContainers` | extra sidecar containers to add to the deployment | `[]`
 | `haproxy.frontendPort` | Port for the Proxy to listen on the frontend | 30636
 | `haproxy.config`       | Suffix of the used config-map | ldap-tls
 | `nodeSelector` | Pod node selector | `{}`
 | `tolerations` | Pod tolerations | `[]`
 | `affinity` | Pod affinity rules | `{}`
+| `resources` | HaProxy resource requests/limits | `{ limits: { cpu: 100m, memory: 128Mi }, requests: { cpu: 100m, memory 128Mi } }`
```
Comment (srueg) [inline haproxy/README.md:60]:
```suggestion
| `resources` | HAProxy resource requests/limits | `{ limits: { cpu: 100m, memory: 128Mi }, requests: { cpu: 100m, memory 128Mi } }`
```

File: haproxy/README.md [hunk 2]
```diff
@@ -94,4 +118,15 @@ Set `haproxy.config` to `galerak8s` to use the Galera configuration with DNS ser
 | `haproxy.galerak8s.nodeCount` | Max number of nodes in the backend | 3
 | `haproxy.galerak8s.port` | Port of the Galera node | 3306
 
+### redisk8s
+
+Set `haproxy.config` to `redisk8s` to use the Redus configuration with DNS service discovery. See https://www.haproxy.com/de/blog/dns-service-discovery-haproxy/
```
Comment (srueg) [inline haproxy/README.md:123]:
```suggestion
Set `haproxy.config` to `redisk8s` to use the Redis configuration with DNS service discovery. See https://www.haproxy.com/de/blog/dns-service-discovery-haproxy/
```

File: haproxy/README.md [hunk 3]
```diff
@@ -94,4 +118,15 @@ Set `haproxy.config` to `galerak8s` to use the Galera configuration with DNS ser
...
+
+| Parameter              | Description            | Default
+|---                     | ---                    | ---
+| `haproxy.redisk8s.check.enabled` | If check should be enabled | true
+| `haproxy.redisk8s.check.redis.auth` | If Authentication credentials for Redis |
```
Comment (srueg) [inline haproxy/README.md:128]: Would be nice to be able to reference an external secret here as well.

File: haproxy/templates/configmap-galera.yaml [hunk 4]
```diff
@@ -22,6 +22,20 @@ data:
...
 
+    {{- if .Values.metrics.enabled }}
+    frontend stats
+      mode http
+      bind *:9000
+
+      stats enable
+      stats uri /stats
+      stats refresh 3s
+      stats admin if TRUE
```
Comment (srueg) [inline haproxy/templates/configmap-galera.yaml:33]:
```suggestion
```
Let's not expose the /stats endpoint as ADMIN here.

File: haproxy/templates/configmap-galerak8s.yaml [hunk 5]
```diff
@@ -28,6 +28,20 @@ data:
...
 
+    {{- if .Values.metrics.enabled }}
+    frontend stats
+      mode http
+      bind *:9000
+
+      stats enable
+      stats uri /stats
+      stats refresh 3s
+      stats admin if TRUE
```
Comment (srueg) [inline haproxy/templates/configmap-galerak8s.yaml:39]:
```suggestion
```

File: haproxy/templates/configmap-ldap-tls.yaml [hunk 6]
```diff
@@ -25,6 +25,20 @@ data:
...
 
+    {{- if .Values.metrics.enabled }}
+    frontend stats
+      mode http
+      bind *:9000
+
+      stats enable
+      stats uri /stats
+      stats refresh 3s
+      stats admin if TRUE
```
Comment (srueg) [inline haproxy/templates/configmap-ldap-tls.yaml:36]:
```suggestion
```

File: haproxy/templates/configmap-redisk8s.yaml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+    {{- if .Values.metrics.enabled }}
+    frontend stats
+      mode http
+      bind *:9000
+
+      stats enable
+      stats uri /stats
+      stats refresh 3s
+      stats admin if TRUE
```
Comment (srueg) [inline haproxy/templates/configmap-redisk8s.yaml:37]:
```suggestion
```

File: haproxy/templates/configmap-redisk8s.yaml [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+      option clitcpka
+      default_backend redis-nodes
+
+    backend redis-nodes
+      mode tcp
+
+      {{ if $redisk8s.check.enabled }}
+      option tcp-check
+      tcp-check connect
+      tcp-check send AUTH\ {{ $redisk8s.check.auth }}\r\n
```
Comment (srueg) [inline haproxy/templates/configmap-redisk8s.yaml:55]: Let's use an environment variable here and mount it from a secret.

Comment (srueg) [reply haproxy/templates/configmap-redisk8s.yaml:55]: Also, AUTH should be optional. If no password/existingSecret is provided, just skip these two lines.

File: haproxy/templates/metrics-svc.yaml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  loadBalancerIP: {{ .Values.metrics.service.loadBalancerIP }}
+  {{- end }}
+  ports:
+    - name: metrics
+      port: 9000
+      targetPort: metrics
+  selector:
+    app.kubernetes.io/name: {{ include "haproxy.name" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+{{- end }}
```
Comment (srueg) [inline haproxy/templates/metrics-svc.yaml:31]:
```suggestion
{{- end }}

```

File: haproxy/templates/prometheusrule.yaml [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+{{- with .Values.metrics.prometheusRule.additionalLabels }}
+{{- toYaml . | nindent 4 }}
+{{- end }}
+spec:
+{{- with .Values.metrics.prometheusRule.rules }}
+  groups:
+    - name: {{ template "haproxy.name" $ }}
+      rules: {{- tpl (toYaml .) $ | nindent 8 }}
+{{- end }}
+{{- end }}
```
Comment (srueg) [inline haproxy/templates/prometheusrule.yaml:25]:
```suggestion
{{- end }}

```

File: haproxy/templates/servicemonitor.yaml [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    {{- end }}
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ template "haproxy.name" . }}
+      app.kubernetes.io/instance: {{ .Release.Name }}
+      app.kubernetes.io/component: "metrics"
+  namespaceSelector:
+    matchNames:
+    - {{ .Release.Namespace }}
+{{- end -}}
```
Comment (srueg) [inline haproxy/templates/servicemonitor.yaml:34]:
```suggestion
{{- end -}}

```

File: haproxy/README.md [hunk 12]
```diff
@@ -48,11 +48,35 @@ create your own configmap and values.
...
 | `tolerations` | Pod tolerations | `[]`
 | `affinity` | Pod affinity rules | `{}`
+| `resources` | HaProxy resource requests/limits | `{ limits: { cpu: 100m, memory: 128Mi }, requests: { cpu: 100m, memory 128Mi } }`
+
+
+### Metrics
+| Parameter              | Description            | Default
+|---                     | ---                    | ---
+| `metrics.enabled`      | Enables /stats and /metrics endpoint in haproxy | true
+| `metrics.serviceMonitor.monitor` | If true, creates a Prometheus Operator ServiceMonitor (also requires `metrics.enabled` to be true) | false
```
Comment (srueg) [inline haproxy/README.md:67]:
```suggestion
| `metrics.serviceMonitor.enabled` | If true, creates a Prometheus Operator ServiceMonitor (also requires `metrics.enabled` to be true) | false
```

File: haproxy/README.md [hunk 13]
```diff
@@ -48,11 +48,35 @@ create your own configmap and values.
...
+| `metrics.serviceMonitor.namespace` | Optional namespace which Prometheus is running in |
+| `metrics.serviceMonitor.interval` | How frequently to scrape metrics (use by default, falling back to Prometheus' default) | 
+| `metrics.serviceMonitor.selector` | Default to kube-prometheus install (CoreOS recommended), but should be set according to Prometheus install | `{ prometheus: kube-prometheus }`
+| `metrics.service.type` | Kubernetes Service type (haproxy metrics) | `ClusterIP`
+| `metrics.service.annotations` | Annotations for the services to monitor (redis master and redis slave service) | `{}`
```
Comment (srueg) [inline haproxy/README.md:72]:
```suggestion
| `metrics.service.annotations` | Annotations for the metrics service | `{}`
```

**Review by srueg — CHANGES_REQUESTED** [review]

File: haproxy/README.md (hunk 3 above)
Comment (splattner) [reply haproxy/README.md:129]: see haproxy.check.existingSecret

File: haproxy/templates/configmap-redisk8s.yaml (hunk 8 above)
Comment (splattner) [reply haproxy/templates/configmap-redisk8s.yaml]: the auth is now only done if a password or existingSecret is provided

File: haproxy/templates/servicemonitor.yaml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+  endpoints:
+  - port: metrics
+    {{- if .Values.metrics.serviceMonitor.interval }}
+    interval: {{ .Values.metrics.serviceMonitor.interval }}
+    {{- end }}
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ template "haproxy.name" . }}
+      app.kubernetes.io/instance: {{ .Release.Name }}
+      app.kubernetes.io/component: "metrics"
```
Comment (srueg) [inline haproxy/templates/servicemonitor.yaml:30]:
```suggestion
```

File: haproxy/templates/secret-check-auth-redisk8s.yaml [hunk 15]
```diff
@@ -0,0 +1,19 @@
+{{- if and (eq .Values.haproxy.config "redisk8s") .Values.haproxy.redisk8s.check.enabled  }}
+{{- if (empty .Values.haproxy.redisk8s.check.existingSecret) }}
```
Comment (srueg) [inline haproxy/templates/secret-check-auth-redisk8s.yaml:2]:
```suggestion
{{- if (empty .Values.haproxy.check.existingSecret) }}
```

If I'm not mistaken?

**Review by srueg — CHANGES_REQUESTED** [review]

File: haproxy/values.yaml [hunk 16]
```diff
@@ -21,14 +21,23 @@ ingress:
...
   type: ClusterIP
+  ## Extra ports to add to the service (e.g. for sidecar containers)
+  additionalPorts: []
 
 haproxy:
   config: ldap-tls
   frontendPort: 30636
 
+  check:
+    existingSecret: "" # Use existing secret with authentication credentials for checks. Must contain `auth-credentials`
```
Comment (srueg) [inline haproxy/values.yaml:39]:
```suggestion
    existingSecret: "" # Use existing secret with authentication credentials for checks
    existingSecretKey: auth-
    credentials
```

File: haproxy/templates/deployment.yaml [hunk 17]
```diff
@@ -23,10 +23,21 @@ spec:
         - name: {{ .Chart.Name }}
           image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
           imagePullPolicy: {{ .Values.image.pullPolicy }}
+          env:
+          - name: CHECK_AUTH_CREDENTIALS
+            valueFrom:
+              secretKeyRef:
+                name: {{ if not (empty .Values.haproxy.check.existingSecret) }}{{ .Values.haproxy.check.existingSecret}}{{ else }}{{ include "haproxy.fullname" . }}-check-auth {{ end }}
+                key: auth-credentials
```
Comment (srueg) [inline haproxy/templates/deployment.yaml:31]:
```suggestion
                key: {{ .Values.haproxy.check.existingSecretKey }}
```

**Review by srueg — CHANGES_REQUESTED** [review]
Let's make the existing secret key configurable to make it easier to reuse secrets (e.g. from another Helm release)

File: haproxy/values.yaml (hunk 16 above)
Comment (splattner) [reply haproxy/values.yaml]: I implemented it a little bit different:

```
  check:
    existingSecret: 
      name: "" # Use existing secret with authentication credentials for checks.
      key: "auth-credentials" 
```

**Review by srueg — APPROVED** [review]

**Pull request merged**